### PR TITLE
Allow filtering output hatches

### PR DESCRIPTION
**Pull request by serenibyss**
Allows for Output Hatches to lock their fluids. Tested a few scenarios:

- 1 Output Hatch locked to the correct fluid for a recipe works properly (runs as normal)
- 1 Output Hatch locked to the incorrect fluid for a recipe works properly (does not run)
- In the above scenario, when the lock is removed, the recipe works properly immediately (starts running, i.e. notifies the controller)
- 2 Output Hatches locked to the correct fluids works properly (runs as normal)

I also tested a more complex scenario:
1. Add 1 Output Hatch, locked to the desired Fluid.
2. Enable "Void Fluids" on the multiblock
3. Run a recipe that outputs 2 Fluids
4. Multiblock will run, outputting only the desired fluid.
5. This can be combined with a Fluid Detector Cover + Machine Controller Cover to automatically void an undesired Fluid, and shut off the multiblock if the desired Fluid's Hatch fills

### CODE REVIEW

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 1]
```diff
@@ -77,6 +88,10 @@ public void readFromNBT(NBTTagCompound data) {
         if (data.hasKey("ContainerInventory")) {
             MetaTileEntityQuantumTank.legacyTankItemHandlerNBTReading(this, data.getCompoundTag("ContainerInventory"), 0, 1);
         }
+        if (isExportHatch && data.hasKey("IsLocked")) {
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:91]: This `hasKey()` check is unneeded. If removed, the `getBoolean("isLocked")` call will return false, and the lockedFluid will be set to null, which does not create any compatibility issues.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 2]
```diff
@@ -180,23 +201,97 @@ protected ModularUI createUI(EntityPlayer entityPlayer) {
...
-                .setHideTooltip(true).setAlwaysShowFull(true);
-        builder.widget(tankWidget);
-        builder.label(11, 20, "gregtech.gui.fluid_amount", 0xFFFFFF);
-        builder.dynamicLabel(11, 30, tankWidget::getFormattedFluidAmount, 0xFFFFFF);
-        builder.dynamicLabel(11, 40, tankWidget::getFluidLocalizedName, 0xFFFFFF);
+        TankWidget tankWidget;
+
+        // Add input/output-specific widgets
+        if (isExportHatch) {
+            tankWidget = new PhantomTankWidget(fluidTank, 69, 43, 18, 18,
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:210]: I think this GUI would be better done similarly to the Quantum Tank gui, where the locked fluid is integrated into the fluid storage slot.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 3]
```diff
@@ -213,4 +308,52 @@ public void addToolUsages(ItemStack stack, @Nullable World world, List<String> t
...
+
+        @Override
+        public boolean canFillFluidType(FluidStack fluid) {
+            if (!isExportHatch) return super.canFillFluidType(fluid);
+            return !locked || lockedFluid == null || fluid.isFluidEqual(lockedFluid);
+        }
+
+        public void onContentsChangedInternal() {
+            this.onContentsChanged();
+        }
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:357]: Can this method be removed, and calls replaced with `onContentsChanged()`? It seems unnecessary right now.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 4]
```diff
@@ -180,23 +201,97 @@ protected ModularUI createUI(EntityPlayer entityPlayer) {
...
+        return (list) -> {
+            String fluidName = "";
+            // If there is no fluid in the tank
+            if (tankWidget.getFluidLocalizedName().isEmpty()) {
+                // But there is a locked fluid
+                if (this.lockedFluid != null) {
+                    fluidName = this.lockedFluid.getLocalizedName();
+                }
+            }
+            else {
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:265]: Formatting: these should be on one line.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 5]
```diff
@@ -180,23 +201,97 @@ protected ModularUI createUI(EntityPlayer entityPlayer) {
...
+            String fluidAmount = "";
+
+            // Nothing in the tank
+            if (tankWidget.getFormattedFluidAmount().equals("0")) {
+                // Display Zero to show information about the locked fluid
+                if (this.lockedFluid != null) {
+                    fluidAmount = "0";
+                }
+            }
+            else {
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:286]: Formatting: these should be on one line.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 6]
```diff
@@ -213,4 +308,53 @@ public void addToolUsages(ItemStack stack, @Nullable World world, List<String> t
...
+
+        @Override
+        public boolean canFillFluidType(FluidStack fluid) {
+            if (!isExportHatch) return super.canFillFluidType(fluid);
+            return !locked || lockedFluid == null || fluid.isFluidEqual(lockedFluid);
+        }
+
+        @Override
+        public void onContentsChanged() {
+            super.onContentsChanged();
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:357]: A comment signifying that this override for widening visibility would be helpful.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java [hunk 7]
```diff
@@ -213,4 +308,53 @@ public void addToolUsages(ItemStack stack, @Nullable World world, List<String> t
...
+
+    private boolean isLocked() {
+        return this.locked;
+    }
+
+    private void setLocked(boolean locked) {
+        if (this.locked == locked) return;
+        this.locked = locked;
+        if (!getWorld().isRemote) {
+            markDirty();
```
Comment (TechLord22) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityFluidHatch.java:320]: Does a `customData` message also need to be sent here?

**Comment by ghzdude** [comment]
I've found an issue where the LCR, while running the ethylbenzene => styrene + hydrogen recipe, voids the output fluids after having the two lv output hatches filtered to styrene (top) and hydrogen (bottom) and then removing the filter on both hatches.

image

Then, trying to use the filter button would indicate that the LCR is trying to output hydrogen to the top output hatch and styrene to the bottom output hatch, but the output hatches are not filled even when they are unfiltered. Filtering one of the hatches to the correct fluid makes the LCR invalidate, and setting both of the hatches to be filtered makes them output fluid correctly.

**Review by ghzdude — APPROVED** [review]

**Pull request merged**